### PR TITLE
feat(customer): add encryption type to customer domain type

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

Adding encryption fields for customer and address

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/customers.rs [hunk 1]
```diff
@@ -56,13 +56,13 @@ pub struct CustomerResponse {
     pub customer_id: String,
     /// The customer's name
     #[schema(max_length = 255, example = "Jon Test")]
-    pub name: Option<String>,
+    pub name: Option<crypto::Encryptable<Secret<String>>>,
     /// The customer's email address
     #[schema(value_type = Option<String>,max_length = 255, example = "JonTest@test.com")]
-    pub email: Option<Secret<String, pii::Email>>,
+    pub email: Option<crypto::Encryptable<Secret<String, pii::Email>>>,
```
Comment (jarnura) [inline crates/api_models/src/customers.rs:62]: Use Type alias like `OptionalEncryptableSecretEmail`

File: crates/api_models/src/customers.rs [hunk 2]
```diff
@@ -56,13 +56,13 @@ pub struct CustomerResponse {
     pub customer_id: String,
     /// The customer's name
     #[schema(max_length = 255, example = "Jon Test")]
-    pub name: Option<String>,
+    pub name: Option<crypto::Encryptable<Secret<String>>>,
```
[the remark this answers is not in the document]
Comment (jarnura) [reply crates/api_models/src/customers.rs:59]: Another Suggestion, we can have type alias like 

`OptionalEncryptableName`
`OptionalEncryptableEmail`
`OptionalEncryptablePhone` etc. Just follow one type throw out the application for name and respective fields

File: crates/router/src/core/customers.rs [hunk 3]
```diff
@@ -34,23 +40,69 @@ pub async fn create_customer(
...
-            line2: customer_address.line2,
-            line3: customer_address.line3,
-            zip: customer_address.zip,
-            state: customer_address.state,
-            first_name: customer_address.first_name,
-            last_name: customer_address.last_name,
-            phone_number: customer_data.phone.clone(),
+            line1: customer_address
+                .line1
+                .async_map(|inner| crypto::Encryptable::encrypt(inner, &key, GcmAes256 {}))
```
Comment (jarnura) [inline crates/router/src/core/customers.rs:59]: create a closure called encrypt = |inner| crypto::Encryptable::encrypt(inner, &key, GcmAes256 {});

so we can do: 
 ``` 
   customer_address
     .line1
     .async_map(encrypt)
```

File: crates/router/src/core/customers.rs [hunk 4]
```diff
@@ -176,17 +243,53 @@ pub async fn delete_customer(
...
-        line3: Some(REDACTED.to_string().into()),
-        state: Some(REDACTED.to_string().into()),
-        zip: Some(REDACTED.to_string().into()),
-        first_name: Some(REDACTED.to_string().into()),
-        last_name: Some(REDACTED.to_string().into()),
-        phone_number: Some(REDACTED.to_string().into()),
+        line1: Some(
+            Encryptable::encrypt(REDACTED.to_string().into(), &key, GcmAes256 {})
+                .await
+                .change_context(errors::ApiErrorResponse::InternalServerError)?,
```
Comment (jarnura) [inline crates/router/src/core/customers.rs:256]: We don't want to handle InternalServerError like this.

you can create a closure which returns Result<X,Y>, Y can be discarded with underscore and just do a change_context once.

```
let update_addressupdate_address = || -> Result<_, _> {
	...
	line1: Some( Encryptable::encrypt(REDACTED.to_string().into(), &key, GcmAes256 {})
                .await?;
    ...
}().change_context(errors::ApiErrorResponse::InternalServerError)
```

Comment (jarnura) [reply crates/router/src/core/customers.rs:256]: Other places also do this change

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/customers.rs (hunk 3 above)
Comment (NishantJoshi00) [reply crates/router/src/core/customers.rs:59]: cool bro 👍

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/types/domain/address.rs [hunk 5]
```diff
@@ -57,24 +68,64 @@ impl behaviour::Conversion for Address {
...
+
+        async {
+            Ok(Self {
+                id: Some(other.id),
+                address_id: other.address_id,
+                city: other.city,
+                country: other.country,
+                line1: other
+                    .line1
+                    .async_map(|inner| Encryptable::decrypt(inner, key, GcmAes256 {}))
```
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/types/domain/address.rs:81]: No, If we were to use closure here, this won't work, as the `decrypt` function returns a generic as the conversion is from `Encryption` to `Encryptable<Secret<T>>`

File: crates/router/src/core/customers.rs [hunk 6]
```diff
@@ -34,48 +40,77 @@ pub async fn create_customer(
...
+                customer_id: customer_id.to_string(),
+                merchant_id: merchant_id.to_string(),
+                id: None,
+                address_id: generate_id(consts::ID_LENGTH, "add"),
+                created_at: common_utils::date_time::now(),
+                modified_at: common_utils::date_time::now(),
+            })
+        }
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)?;
```
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/core/customers.rs:90]: Cool!

**Review by jarnura — APPROVED** [review]

**Pull request merged**